pkg/storage/external/client.go

```diff
@@ -0,0 +1,185 @@
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	resp, err := s.c.Do(req)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return nil, errors.E(op, fmt.Errorf("unexpected status code: %d", resp.StatusCode), resp.StatusCode)
+	}
+	list := []string{}
+	scnr := bufio.NewScanner(resp.Body)
+	for scnr.Scan() {
+		list = append(list, scnr.Text())
+	}
+	if scnr.Err() != nil {
+		return nil, errors.E(op, scnr.Err())
+	}
+	return list, nil
+}
+
+func (s *service) Info(ctx context.Context, module, vsn string) ([]byte, error) {
+	const op errors.Op = "external.Info"
+	url := s.url + "/" + module + "/@v/" + vsn + ".info"
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	resp, err := s.c.Do(req)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return nil, errors.E(op, fmt.Errorf("none 200 status code: %v", resp.StatusCode), resp.StatusCode)
+	}
+	info, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	return info, nil
+}
+
+func (s *service) GoMod(ctx context.Context, module, vsn string) ([]byte, error) {
+	const op errors.Op = "external.GoMod"
+	url := s.url + "/" + module + "/@v/" + vsn + ".mod"
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	resp, err := s.c.Do(req)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return nil, errors.E(op, fmt.Errorf("none 200 status code: %v", resp.StatusCode), resp.StatusCode)
+	}
+	info, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	return info, nil
+}
+
+func (s *service) Zip(ctx context.Context, module, vsn string) (io.ReadCloser, error) {
+	const op errors.Op = "external.Zip"
+	url := s.url + "/" + module + "/@v/" + vsn + ".zip"
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	resp, err := s.c.Do(req)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	if resp.StatusCode != 200 {
+		return nil, errors.E(op, fmt.Errorf("none 200 status code: %v", resp.StatusCode), resp.StatusCode)
+	}
+	return resp.Body, nil
+}
+
+func (s *service) Exists(ctx context.Context, module, version string) (bool, error) {
+	const op errors.Op = "external.Exists"
+	_, err := s.Info(ctx, module, version)
+	if err != nil {
+		if errors.Is(err, errors.KindNotFound) {
+			err = nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (s *service) Save(ctx context.Context, module, version string, mod []byte, zip io.Reader, info []byte) error {
+	const op errors.Op = "external.Save"
+	url := s.url + "/" + module + "/@v/" + version + ".save"
```

Comment:
why not make this a `POST /module/@v/version`? same with deleting below when it gets implemented (`DELETE /module/@v/version`)

Reply:
@arschles i think it's better to be explicit and to take the `.ext` pattern. For example the download protocol explicitly does `GET /module/@v/list` and not just `GET /module` -- and